packages/playground/blueprints/src/lib/compile.ts

```diff
@@ -116,33 +145,77 @@ export function compileBlueprint(
 			),
 		},
 		run: async (playground: UniversalPHP) => {
+			const progressMap = new Map<StepDefinition, Promise<void>>();
+			const progressPromises = new Map<StepDefinition, PromiseExecutor>();
+			for (const { step } of compiled) {
+				progressMap.set(
+					step,
+					new Promise((resolve, reject) => {
+						progressPromises.set(step, { resolve, reject });
+					})
+				);
+			}
+			onRunningSteps(progressMap);
+
 			try {
 				// Start resolving resources early
-				for (const { resources } of compiled) {
+				for (const { resources, step } of compiled) {
+					const stepPromise = progressPromises.get(
+						step
+					) as PromiseExecutor;
```

Comment:
this `PromiseExecutor` shouldn't be necessary. was it telling you it could be `undefined`? If that's the case, we might be able to try `progressPromises.get(step)!`